### PR TITLE
Finalize Controllers

### PR DESCRIPTION
**Pull request by snaselj**
# Closes Nan

## What's Changed

- Fixed testing `RelatedManager` fields. 
- Added Controller default statuses.
- Added `Devices` section to `Controller` and `ControllerDeviceGroup` retrieve form.
- Changed Controller Device Group list to tree.
- Added `Devices` field to `ControllerDeviceGroup` edit form.
- Added validators.
- Added signal updating controller for all descendants of `ControllerDeviceGroup` on update.

### CODE REVIEW

File: nautobot/dcim/forms.py [hunk 1]
```diff
@@ -4224,11 +4224,14 @@ class Meta:
 class ControllerDeviceGroupForm(NautobotModelForm):
     """ControllerDeviceGroup create/edit form."""
 
+    devices = DynamicModelMultipleChoiceField(queryset=Device.objects.all(), required=False)
```
Comment (snaselj) [inline nautobot/dcim/forms.py:4227]: Doesn't update devices.

Any clue how to fix this or link to the same pattern pls.? @whitej6 @glennmatthews

File: nautobot/core/management/commands/generate_test_data.py [hunk 2]
```diff
@@ -292,7 +292,7 @@ def handle(self, *args, **options):
             self.stdout.write(
                 self.style.WARNING(f'Flushing all existing data from the database "{options["database"]}"...')
             )
-            call_command("flush", "--no-input", "--database", options["database"])
+            call_command("flush", "--no-input", "--database", options["database"], inhibit_post_migrate=True)
```
Comment (glennmatthews) [inline nautobot/core/management/commands/generate_test_data.py:295]: Not familiar with `inhibit_post_migrate` - what's it do, why's it needed? :-)

Comment (snaselj) [reply nautobot/core/management/commands/generate_test_data.py:295]: Before merging main PR into next, I'll separate fixes not directly related to controller models into distinct PRs :-)

File: nautobot/dcim/forms.py (hunk 1 above)
Comment (snaselj) [reply nautobot/dcim/forms.py:4227]: Works nice, TY!

File: nautobot/dcim/models/devices.py [hunk 3]
```diff
@@ -1342,3 +1354,20 @@ class Meta:
...
     def __str__(self):
         return self.name or super().__str__()
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._original_controller = self.controller if self.present_in_database else None
+        self._original_parent = self.parent if self.present_in_database else None
+
+    def clean(self):
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/dcim/models/devices.py:1364]: Seems like a nice-to-have to me.

**Comment by glennmatthews** [comment]
Test failure looks like it's not finding any suitable `Role` to attach to a newly created `Controller`.

File: nautobot/dcim/models/devices.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/models/devices.py:1364]: Added as a todo point to the base PR

**Comment by snaselj** [comment]
> Test failure looks like it's not finding any suitable `Role` to attach to a newly created `Controller`.

Fixed, now I'm solving another issue with `ControllerDeviceGroup` factory. Not sure how to set the `controller` based on the parent group.

**Comment by glennmatthews** [comment]
> > Test failure looks like it's not finding any suitable `Role` to attach to a newly created `Controller`.
> 
> Fixed, now I'm solving another issue with `ControllerDeviceGroup` factory. Not sure how to set the `controller` based on the parent group.

Possibly something along the lines of https://github.com/nautobot/nautobot/blob/develop/nautobot/dcim/factory.py#L457-L477?

File: nautobot/core/management/commands/generate_test_data.py (hunk 2 above)
Comment (snaselj) [reply nautobot/core/management/commands/generate_test_data.py]: Removed this fix, as it breaks the current tests.

File: nautobot/dcim/signals.py [hunk 4]
```diff
@@ -286,3 +287,21 @@ def prevent_adding_tagged_vlans_with_incorrect_mode_or_site(sender, instance, ac
...
+def handle_controller_device_group_controller_change(instance, created, raw=False, **kwargs):
+    """Update descendants when the controller changes."""
+    if raw or created:
+        return
+
+    with transaction.atomic():
+        if instance._original_controller != instance.controller:
+            for group in instance.descendants(include_self=False):
+                group.controller = instance.controller
+                group.save()
```
Comment (glennmatthews) [inline nautobot/dcim/signals.py:307]: Maybe `.validated_save()`? Not blocking.

File: nautobot/docs/user-guide/core-data-model/dcim/controllerdevicegroup.md [hunk 5]
```diff
@@ -1,3 +1,13 @@
 # Controller Device Groups
 
-TBD
+A Controller Device Group in Nautobot models the relationship between controllers and the devices they manage. It organizes devices under specific controllers, providing a structured way to represent devices controlled by a central network management system. These systems might include, but are not limited to, software-defined networking (SDN) controllers, wireless controllers, and cloud-based network management platforms.
```
Comment (glennmatthews) [inline nautobot/docs/user-guide/core-data-model/dcim/controllerdevicegroup.md:3]: Cross-link to `controller` and `device` documentation?

File: nautobot/docs/user-guide/core-data-model/dcim/controller.md [hunk 6]
```diff
@@ -1,3 +1,13 @@
 # Controllers
 
-TBD
+A Controller in Nautobot is an abstraction meant to represent network or SDN (Software-Defined Networking) controllers. These may include, but are not limited to, wireless controllers, cloud-based network management systems, and other forms of central network control mechanisms.
+
+The Controller model is central to representing an entity that oversees one or more devices or device groups. It serves as the focal point for control and management operations within a network. A Controller can be associated with either individual devices or with groups of devices through a `DeviceRedundancyGroup`.
```
Comment (glennmatthews) [inline nautobot/docs/user-guide/core-data-model/dcim/controller.md:5]: Cross-link to `device` and `DeviceRedundancyGroup` documentation?

**Review by glennmatthews — APPROVED** [review]
Let's merge this down to get the full branch ready for re-review. A few more comments but nothing blocking.

**Pull request merged**